src/ucp/core/ucp_am.c

```diff
@@ -968,13 +970,21 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nbx,
     const ucp_request_send_proto_t *proto;
     size_t contig_length;
     ucp_operation_id_t op_id;
+    ucs_memory_type_t mem_type;
 
     UCP_CONTEXT_CHECK_FEATURE_FLAGS(worker->context, UCP_FEATURE_AM,
                                     return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
     UCP_REQUEST_CHECK_PARAM(param);
 
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
 
+    if (ucs_likely(ucs_memtype_cache_is_empty())) {
+        mem_type = UCS_MEMORY_TYPE_HOST;
+    } else {
+        mem_type = (param->op_attr_mask & UCP_OP_ATTR_FIELD_MEMORY_TYPE) ?
+                    param->memory_type : UCS_MEMORY_TYPE_UNKNOWN;
```

Comment:
IMO we can still avoid the extra branch when memtype cache is empty. since after this PR there would be 2 branches on the ucs_memtype_cache_is_empty() check instead of one.

Reply:
How about I initialize mem_type to `UCS_MEMORY_TYPE_HOST` and update it `if (!ucs_memtype_cache_is_empty())` ?

Reply:
same effect. need to push checking mem_type bit field_mask after the existing check of ucs_memtype_cache_is_empty

Reply:
that means something like:
```diff
diff --git a/src/ucp/proto/proto_am.inl b/src/ucp/proto/proto_am.inl
index cdd867120b..2c1f5e93ce 100644
--- a/src/ucp/proto/proto_am.inl
+++ b/src/ucp/proto/proto_am.inl
@@ -596,11 +596,13 @@ ucp_proto_get_short_max(const ucp_request_t *req,
 
 static UCS_F_ALWAYS_INLINE int
 ucp_proto_is_inline(ucp_ep_h ep, const ucp_memtype_thresh_t *max_eager_short,
-                    ssize_t length)
+                    ssize_t length, const ucp_request_param_t *param)
 {
     return (ucs_likely(length <= max_eager_short->memtype_off) ||
             (length <= max_eager_short->memtype_on &&
-             ucs_memtype_cache_is_empty()));
+             ucs_memtype_cache_is_empty())) ||
+           ((param->op_attr_mask & UCP_OP_ATTR_FIELD_MEMORY_TYPE) &&
+            (param->memory_type == UCS_MEMORY_TYPE_HOST));
 }
```